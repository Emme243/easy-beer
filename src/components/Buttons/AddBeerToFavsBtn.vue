<template>
  <v-tooltip bottom color="deep-purple" class="d-flex align-center">
    <template #activator="{ on, attrs }">
      <button v-bind="attrs" v-on="on" @click="handleBeerInFavorites">
        <v-icon :color="iconColor" size="20">{{ heartIcon }}</v-icon>
      </button>
    </template>
    <v-icon dark small>{{ heartIcon }}</v-icon>
    <span class="ml-2">{{ tooltipText }}</span>
  </v-tooltip>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';

export default {
  name: 'AddBeerToFavsBtn',
  props: {
    beer: {
      type: Object,
      required: true,
    },
  },
  methods: {
    ...mapActions({
      addBeerToFavoritesInStore: 'favoriteBeers/addBeerToFavorites',
      removeBeerFromFavoritesInStore: 'favoriteBeers/removeBeerFromFavorites',
    }),
    handleBeerInFavorites() {
      if (this.isBeerInFavorites) this.removeBeerFromFavoritesInStore(this.beer.id);
      else this.addBeerToFavoritesInStore(this.beer);
    },
  },
  computed: {
    ...mapGetters({
      isBeerInFavoritesFromStore: 'favoriteBeers/isBeerInFavorites',
    }),
    isBeerInFavorites() {
      return this.isBeerInFavoritesFromStore(this.beer.id);
    },
    heartIcon() {
      return this.isBeerInFavorites ? 'fa fa-heart-circle-check' : 'fa fa-heart-circle-plus';
    },
    iconColor() {
      return this.isBeerInFavorites ? 'deep-purple lighten-4' : 'white';
    },
    tooltipText() {
      return this.isBeerInFavorites ? 'Cerveza en favoritos' : 'Agregar a favoritos';
    },
  },
};
</script>

<style scoped></style>
