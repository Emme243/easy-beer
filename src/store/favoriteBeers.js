export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
