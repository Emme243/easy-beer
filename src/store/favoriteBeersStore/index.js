export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
