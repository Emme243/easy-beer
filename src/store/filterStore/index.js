import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
