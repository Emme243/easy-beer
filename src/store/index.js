import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
