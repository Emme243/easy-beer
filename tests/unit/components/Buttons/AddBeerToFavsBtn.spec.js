import AddBeerToFavsBtn from '@/components/Buttons/AddBeerToFavsBtn';
import store from '@/store';
import { createLocalVue, mount } from '@vue/test-utils';
import Vuetify from 'vuetify';
import { beersMock } from '../../mocks/beersMock';

describe('Pruebas en AddBeerToFavsBtn.vue', function () {
  let vuetify;
  const localVue = createLocalVue();

  const beer = beersMock[0];
  const mountFunction = options => {
    return mount(AddBeerToFavsBtn, {
      localVue,
      vuetify,
      store,
      propsData: { beer },
      ...options,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const heartIconBtnSelector = '[data-test-id="heart-icon-button"]';
  const heartIconSelector = '[data-test-id="heart-icon"]';

  it('Debe hacer match con el snapshot', () => {
    const wrapper = mountFunction();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('El botón debe tener el icono de corazón debe tener la clase `fa-heart-circle-plus` cuando la cerveza no está en favoritos', () => {
    const wrapper = mountFunction();
    const heartIconElement = wrapper.find(heartIconSelector);
    expect(heartIconElement.exists()).toBe(true);
    expect(heartIconElement.classes('fa-heart-circle-plus')).toBe(true);
  });

  it('El botón debe tener el icono de corazón debe tener la clase `fa-heart-circle-check` cuando la cerveza está en favoritos', async () => {
    const wrapper = mountFunction();
    const heartIconElement = wrapper.find(heartIconSelector);
    expect(heartIconElement.exists()).toBe(true);
    const heartIconBtn = wrapper.find(heartIconBtnSelector);
    expect(heartIconBtn.exists()).toBe(true);
    await heartIconBtn.trigger('click');
    expect(heartIconElement.classes('fa-heart-circle-check')).toBe(true);
  });

  it('El botón con el ícono de corazón para añadir o remover la cerveza a favoritos debe llamar al dispatch handleBeerInFavorites con la cerveza', async () => {
    store.dispatch = jest.fn();
    const wrapper = mountFunction();
    const heartIconBtn = wrapper.find(heartIconBtnSelector);
    expect(heartIconBtn.exists()).toBe(true);
    await heartIconBtn.trigger('click');
    expect(store.dispatch).toHaveBeenCalledWith('favoriteBeers/handleBeerInFavorites', beer);
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
