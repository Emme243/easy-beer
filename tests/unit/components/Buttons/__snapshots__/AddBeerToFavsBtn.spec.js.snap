// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddBeerToFavsBtn.vue Debe hacer match con el snapshot 1`] = `
<span
  class="v-tooltip d-flex align-center v-tooltip--bottom"
>
  <!---->
  <button
    aria-expanded="false"
    aria-haspopup="true"
    data-test-id="heart-icon-button"
  >
    <i
      aria-hidden="true"
      class="v-icon notranslate fa fa fa fa-heart-circle-plus theme--light white--text"
      data-test-id="heart-icon"
      style="font-size: 20px;"
    />
  </button>
</span>
`;
